File: [UVA]/10189.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

vector<pair<int, int>> getAdj(int i, int j, int m, int n)
{
    vector<pair<int, int>> vizinhos;
    int dx[] = {-1, -1, -1, 0, 0, 1, 1, 1};
    int dy[] = {-1, 0, 1, -1, 1, -1, 0, 1};

    for (int d = 0; d < 8; d++)
    {
        int x = i + dx[d];
        int y = j + dy[d];
        if (x >= 0 && x < m && y >= 0 && y < n)
        {
            vizinhos.push_back({x, y});
        }
    }

    return vizinhos;
}

int main()
{
    int m, n;
    int jogos = 0;
    vector<string> resultados;

    while (true)
    {
        cin >> m >> n;
        if (m == 0 && n == 0)
            break;
        jogos++;

        vector<vector<int>> campo(m, vector<int>(n, 0));
        vector<vector<char>> dicas(m, vector<char>(n));

        for (int i = 0; i < m; i++)
        {
            string linha;
            cin >> linha;
            for (int j = 0; j < n; j++)
            {
                dicas[i][j] = linha[j];
            }
        }

        for (int i = 0; i < m; i++)
        {
            for (int j = 0; j < n; j++)
            {
                if (dicas[i][j] == '*')
                {
                    campo[i][j] = -1;
                    for (auto [x, y] : getAdj(i, j, m, n))
                    {
                        if (campo[x][y] != -1)
                            campo[x][y]++;
                    }
                }
            }
        }

        string resultado = "Field #" + to_string(jogos) + ":\n";
        for (int i = 0; i < m; i++)
        {
            for (int j = 0; j < n; j++)
            {
                if (campo[i][j] == -1)
                    resultado += '*';
                else
                    resultado += char(campo[i][j] + '0');
            }
            resultado += '\n';
        }

        resultados.push_back(resultado);
    }

    for (int i = 0; i < (int)resultados.size(); i++)
    {
        cout << resultados[i];
        if (i != (int)resultados.size() - 1)
            cout << '\n';
    }

    return 0;
}

File: [UVA]/10189.py
jogos = 0
resultados = []

def getAdj(i,j,m,n):
    direcoes = [
        (-1, -1), (-1, 0), (-1, 1),
        (0, -1),           (0, 1),
        (1, -1),  (1, 0),  (1, 1)
    ]
    vizinhos = []
    for x,y in direcoes:
        x+=i
        y+=j
        if (x>=0 and x<m and y>=0 and y<n):
            vizinhos.append((x,y))
    return vizinhos


while (True):
    m,n = list(map(int, input().split()))
    if (m == 0 and n == 0):break

    jogos+=1

    campo = [[0]*n for _ in range(m)]
    dicas = []
    
    for _ in range(m):
        linha = list(input())
        dicas.append(linha)

    for i in range(m):
        for j in range(n):
            if (dicas[i][j]) == "*":
                campo[i][j] = -1
                v = getAdj(i,j,m,n)
                for x,y in v:
                    if campo[x][y] != -1:
                        campo[x][y]+=1

    saida = [f"Field #{jogos}:"]
    for i in campo:
        linha = ''.join('*' if j == -1 else str(j) for j in i)
        saida.append(linha)
    resultados.append('\n'.join(saida))

print('\n\n'.join(resultados))
